packages/remirror__extension-file/src/file-extension.tsx

```diff
@@ -0,0 +1,355 @@
+import {
+  findPlaceholderPlayload,
+  findPlaceholderPos,
+  placeholderPlugin,
+  setPlaceholderAction,
+} from './file-placeholder-plugin';
+
+export interface FileOptions {
+  loadFileHandler?: LoadFileHandler;
+  getAttrsBeforeUpload?: GetAttrsBeforeUpload;
+  getAttrsAfterUpload?: GetAttrsAfterUpload;
+  render?: (props: FileComponentProps) => React.ReactElement<HTMLElement> | null;
+}
+
+@extension<FileOptions>({
+  defaultOptions: { ...defaultOptions, render: FileComponent },
+})
+export class FileExtension extends NodeExtension<FileOptions> {
+  get name() {
+    return 'file' as const;
+  }
+
+  ReactComponent: ComponentType<NodeViewComponentProps> = (props) => {
+    const fileLoader: FileLoader | undefined = findPlaceholderPlayload(
+      props.view.state,
+      props.node.attrs.id,
+    );
+    return this.options.render({ ...props, fileLoader });
+  };
+
+  createTags() {
+    return [ExtensionTag.Block];
+  }
+
+  createNodeSpec(extra: ApplySchemaAttributes, override: NodeSpecOverride): NodeExtensionSpec {
+    return {
+      attrs: {
+        id: { default: null },
+        ...extra.defaults(),
+        url: { default: '' },
+        fileName: { default: '' },
+        fileType: { default: '' },
+        fileSize: { default: 0 },
+        error: { default: null },
+      },
+      ...override,
+      selectable: false,
+      draggable: true,
+      atom: true,
+      content: '',
+      parseDOM: [
+        {
+          tag: '[data-file]',
+          priority: ExtensionPriority.Low,
+          getAttrs: (dom) => {
+            const anchor = dom as HTMLAnchorElement;
+            const url = anchor.getAttribute('href');
+            const fileName = anchor.getAttribute('data-filename');
+            const fileType = anchor.getAttribute('data-filetype');
+            const fileSize = anchor.getAttribute('data-filesize');
+
+            return {
+              ...extra.parse(dom),
+              url,
+              fileName,
+              fileType,
+              fileSize,
+            };
+          },
+        },
+      ],
+      toDOM: (node) => {
+        const { url, ...rest } = omitExtraAttributes(node.attrs, extra);
+        const attrs = {
+          ...extra.dom(node),
+          ...rest,
+          href: url,
+          'data-file': '',
+          'data-filename': node.attrs.fileName,
+          'data-filetype': node.attrs.fileType,
+          'data-filesize': node.attrs.fileSize,
+        };
+
+        return ['a', attrs];
+      },
+    };
+  }
+
+  createExternalPlugins(): ProsemirrorPlugin[] {
+    return [placeholderPlugin];
+  }
+
+  createPasteRules(): PasteRule[] {
+    return [
+      {
+        type: 'file',
+        fileHandler: (props) => {
+          let pos: number | undefined;
+
+          if (props.type === 'drop') {
+            pos = props.pos;
+          }
+
+          for (const file of props.files) {
+            this.startLoadFile(file, pos);
```

Comment:
I think we should add an option to this extension that allows you to filter out certain file types. For instance you may wise to ignore images file drops, as they will be handled by the image extension.

Reply:
added a new option `pasteRuleRegexp`